test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java

```diff
@@ -177,6 +182,66 @@ public void initClient() throws IOException {
         assert hasXPack != null;
         assert nodeVersions != null;
     }
+    
+    // Helper class to check warnings in REST responses with sensitivity to versions
+    // used in the target cluster.
+    protected class VersionSensitiveWarningsHandler implements WarningsHandler {
+        Set<String> requiredSameVersionClusterWarnings = new HashSet<>();
+        Set<String> allowedWarnings = new HashSet<>();
+
+        public void current(String... requiredWarnings) {
+            requiredSameVersionClusterWarnings.addAll(Arrays.asList(requiredWarnings));
+            if(clusterHasACurrentVersionNode()) {
+                // as a convenience add all of the current-version warnings to the allowed
+                // warnings because at least one node in the cluster is a current-version node
+                // and so we assume it can return these responses
+                allowedWarnings.addAll(Arrays.asList(requiredWarnings));
+            }
+        }
+
+        public void other(String... allowedWarnings) {            
+            this.allowedWarnings.addAll(Arrays.asList(allowedWarnings));
+        }
+
+        @Override
+        public boolean warningsShouldFailRequest(List<String> warnings) {
+            if (isExclusivelyTargetingCurrentVersionCluster()) {
+                // absolute equality required in expected and actual.
+                Set<String> actual = new HashSet<>(warnings);
+                return false == requiredSameVersionClusterWarnings.equals(actual);
+            } else {
+                // Some known warnings can safely be ignored
+                for (String actualWarning : warnings) {
+                    if (false == allowedWarnings.contains(actualWarning)) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+        }
+        
+        private boolean isExclusivelyTargetingCurrentVersionCluster() {
+            return nodeVersions.size() == 0 && nodeVersions.first().equals(Version.CURRENT);
+        }
+        
+        private boolean clusterHasACurrentVersionNode() {
+            for (Version version : nodeVersions) {
+                if(version.equals(Version.CURRENT)) {
+                    return true;
+                }                
+            }
+            return false;
+        }    
+        
+    }
+    
+    protected RequestOptions expectVersionSpecificWarnings(Consumer<VersionSensitiveWarningsHandler> expectationsSetter) {
```

Comment:
Same argument for `public` and `static`.

Do you think this should take `VersionSensistiveWarningsHandler` as an argument?

Reply:
The main goal of this helper method was trying to minimise the code required in tests to pass warning lists (I suspect we'll end up with a lot of calls from tests).
I wanted tests to call a helper function with `String...` typed params but of course a method/constructor can only take one of those variable length params which is why I needed 2 method invocations and this seemed like a shorthand way of doing that without clients having to call ` expect(new VersionSensitiveWarningsHandler(currentWarnings).other(allowedWarnings))`.  We'd also need to pass `Set<Version> nodeVersions` into this object too. Happy to take any alternative suggestions though.

Reply:
Hmmm. I see what you mean here. I don't think you need to pass the `Set<Version>` because I think that can be publicly exposed on `ESRestTestCase`. The rest, yeah, I was hoping for a fluent interface, but i see how repeating the words doesn't help. I think taking the `Consumer` is fine. It is certainly easier to read and we're going to have this all over the place. So +1 for keeping the consumer.

Reply:
If you want VersionSensitiveWarningsHandler to work outside of EsRestTestCase (which I think you said was the case) then maybe we make it a standalone class with minimal dependencies and make the `EsRestTestCase.expectVersionSpecificWarnings` impl call VSWH constructor with `Set<Version>` rather than have VSWH reaching back to ESRestTestCase.nodeVersions to see this info

Reply:
I don't think we'd want it to work if you aren't *running* an `ESRestTestCase`, but it'd be nice if it worked if you didn't *extend* from `ESRestTestCase`.